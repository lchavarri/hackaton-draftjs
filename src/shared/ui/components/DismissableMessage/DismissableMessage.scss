@import '../../assets/variables';
@import '../../assets/mixins';

$message-height: rem(32px);
.ui.dismissable.message {
  &.negative {
    background-color: $error-red;
    border-color: $error-red;
    color: $red;
  }
  background-color: $secondary-background-color;
  border-color: $secondary-background-color;
  box-shadow: none;
  border-radius: 0;
  padding: rem(6px 40px 6px 16px);
  margin-top: 0;
  margin-bottom: rem(1px);
  min-height: $message-height;
  width: 100%;
  color: $text;
  p {
    &.clamped {
      @include line-clamp(3);
    }
    font-size: rem(12px);
    font-weight: 300;
    line-height: rem(20px);
    word-break: break-word;
  }
  .icon.close {
    cursor: pointer;
    color: $text;
    display: flex;
    justify-content: center;
    align-items: center;
    top: 0;
    right: rem(4px);
    width: rem(32px);
    height: rem(32px);
    font-size: rem(12px);
    &:before {
      content: '';
    }
    svg {
      width: rem(14px);
      height: rem(14px);
    }
  }
}
