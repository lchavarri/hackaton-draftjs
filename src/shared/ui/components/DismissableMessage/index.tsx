import React, { useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/pro-light-svg-icons';

import './DismissableMessage.scss';

export default function DismissableMessage(props: any) {
  const [visible, setVisible] = useState(true);
  const [clamped, setClamped] = useState(true);
  const { children, header, onClose, variant } = props;

  const handleClose = () => {
    setVisible(false);
    if (onClose) {
      onClose();
    }
  };

  if (!visible) {
    return null;
  }

  return (
    <div className={`ui dismissable message ${variant ? variant : ''}`}>
      <i className="icon close" onClick={handleClose}>
        <FontAwesomeIcon icon={faTimes} />
      </i>
      <div className="content" onClick={() => setClamped(!clamped)}>
        {header ? <div className="header">{header}</div> : <></>}
        <p className={clamped ? 'clamped' : ''}>{children}</p>
      </div>
    </div>
  );
}
