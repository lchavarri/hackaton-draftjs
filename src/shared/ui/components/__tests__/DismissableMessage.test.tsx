import React from 'react';
import { shallow, ReactWrapper } from 'enzyme';
import DismissableMessage from '../DismissableMessage';

describe('DismissableMessage component', () => {
  let component: ReactWrapper;
  const onCloseSpy = jest.fn();
  const props = {
    variant: 'negative',
    header: 'This is a header',
    onClose: null
  };
  const getComponent = () => <DismissableMessage {...props} />;
  const wrap = () => shallow(getComponent());

  beforeEach(() => {
    component = wrap();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match component snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should have variant class', () => {
    expect(component.exists(`.negative`)).toBe(true);
  });

  it('should not have variant class if not provided', () => {
    component.setProps({ variant: null });
    component.update();
    expect(component.exists(`.negative`)).toBe(false);
  });

  describe('when click on close', () => {
    beforeEach(() => {
      component = wrap();
      component.find('.icon.close').simulate('click');
    });

    it('should return null at render phase', () => {
      expect(component.getElement()).toBe(null);
    });

    it('should not call onClose if not provided', () => {
      expect(onCloseSpy).not.toHaveBeenCalled();
    });

    it('should call onClose callback if any', () => {
      component = wrap();
      component.setProps({ onClose: onCloseSpy });
      component.update();
      component.find('.icon.close').simulate('click');
      expect(onCloseSpy).toHaveBeenCalled();
    });
  });

  it('should have clamped class by default', () => {
    expect(component.exists('.clamped')).toBe(true);
  });

  it('should toggle clamped class on click', () => {
    component.find('.content').simulate('click');
    expect(component.exists('.clamped')).toBe(false);
    component.find('.content').simulate('click');
    expect(component.exists('.clamped')).toBe(true);
  });

  it('should show header if provided', () => {
    const headerElm = component.find('.header');
    expect(headerElm.getElement()).toBeTruthy();
    expect(headerElm.html()).toContain(props.header);
  });

  it('should not have header if not provided', () => {
    component.setProps({ header: null });
    component.update();
    expect(component.exists('.header')).toBe(false);
  });
});
