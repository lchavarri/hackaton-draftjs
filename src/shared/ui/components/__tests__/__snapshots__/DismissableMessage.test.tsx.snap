// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DismissableMessage component should match component snapshot 1`] = `
<div
  className="ui dismissable message negative"
>
  <i
    className="icon close"
    onClick={[Function]}
  >
    <mockConstructor
      icon={[MockFunction]}
    />
  </i>
  <div
    className="content"
    onClick={[Function]}
  >
    <div
      className="header"
    >
      This is a header
    </div>
    <p
      className="clamped"
    />
  </div>
</div>
`;
